src/read/pe/resource.rs

```diff
@@ -0,0 +1,527 @@
+use crate::endian::{LittleEndian as LE, U16};
+use crate::pe::{ImageResourceDataEntry, ImageResourceDirectory, ImageResourceDirectoryEntry};
+use crate::read::{
+    pe::{ImageNtHeaders, PeFile},
+    ReadError, ReadRef, Result,
+};
+
+/// A resource directory
+#[derive(Clone, Copy)]
+pub struct ResourceDirectoryTable<'data> {
+    directory_data: &'data [u8],
+    /// the resource directory table
+    pub table: &'data ImageResourceDirectory,
+    /// the resource directory entries
+    pub entries: &'data [ImageResourceDirectoryEntry],
+}
+
+impl<'data> core::fmt::Debug for ResourceDirectoryTable<'data> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        // Do not print self.data
+        f.debug_struct("ResourceDirectoryTable")
+            .field("table", &self.table)
+            .field("entries", &self.entries)
+            .finish()
+    }
+}
+
+impl<'data> ResourceDirectoryTable<'data> {
+    /// Returns the root resource directory, if any.
+    pub fn parse<Pe, R>(pe_file: &PeFile<'data, Pe, R>) -> Result<Self>
+    where
+        Pe: ImageNtHeaders,
+        R: ReadRef<'data>,
+    {
+        let dir = pe_file
+            .data_directory(crate::pe::IMAGE_DIRECTORY_ENTRY_RESOURCE as _)
+            .read_error("No resource directory")?;
+        let data = dir.data(pe_file.data(), &pe_file.section_table())?;
+        Self::read(data, 0)
+    }
+
+    fn read(data: &'data [u8], offset: u64) -> Result<Self> {
+        let table = data
+            .read_at::<ImageResourceDirectory>(offset as _)
```

Comment:
Change this to `read` instead of doing the offset calculation below.

Reply:
Done.

But: in `ReadRef::read_bytes(..)` we do `*offset = offset.wrapping_add(size);` which could end-up reading the next data at the start of the buffer instead of failing (in the other end, it should not fail if are reading exactly the last part of the buffer before overflowing the offset...). I do recognize it's an unlikely possibility.

Reply:
That can't wrap because if it would overflow then `read_bytes_at` would fail first. The `wrapping_add` is only to ensure no panic handling is ever generated, not because it might wrap.